src/frontend/Typechecker.ml

```diff
@@ -391,22 +391,25 @@ let inferred_unsizedtype_of_indexed ~loc ut indices =
     | (UMatrix | UComplexMatrix), [`Multi; `Single] -> vec
     | (UMatrix | UComplexMatrix), _ :: _ :: _ :: _
      |(UVector | URowVector | UComplexRowVector | UComplexVector), _ :: _ :: _
-     |(UInt | UReal | UComplex | UFun _ | UMathLibraryFunction), _ :: _ ->
+     |( (UInt | UReal | UComplex | UFun _ | UMathLibraryFunction | UTuple _)
+      , _ :: _ ) ->
         Semantic_error.not_indexable loc ut (List.length indices) |> error in
   aux ut (List.map ~f:indexing_type indices)
 
-let inferred_ad_type_of_indexed at uindices =
-  UnsizedType.lub_ad_type
-    ( at
-    :: List.map
-         ~f:(function
-           | All -> UnsizedType.DataOnly
-           | Single ue1 | Upfrom ue1 | Downfrom ue1 ->
-               UnsizedType.lub_ad_type [at; ue1.emeta.ad_level]
-           | Between (ue1, ue2) ->
-               UnsizedType.lub_ad_type
-                 [at; ue1.emeta.ad_level; ue2.emeta.ad_level] )
-         uindices )
+let inferred_ad_type_of_indexed at ut uindices =
+  UnsizedType.fill_adtype_for_type
+    (UnsizedType.lub_ad_type
+       ( at
+       :: List.map
+            ~f:(function
+              | All -> UnsizedType.DataOnly
+              | Single ue1 | Upfrom ue1 | Downfrom ue1 ->
+                  UnsizedType.lub_ad_type [at; ue1.emeta.ad_level]
+              | Between (ue1, ue2) ->
+                  UnsizedType.lub_ad_type
+                    [at; ue1.emeta.ad_level; ue2.emeta.ad_level] )
+            uindices ) )
+    ut
```

Comment:
Ok, this is something that has confused me for quite a while: the indices are always integers so they're not differentiable and can be ignored for the purposes of computiing `ad_type`, right? So what's all this complicated logic doing here?
And a new question: tuples can't be indexed so why would this PR need to change anything here?

By the way, I don't think `UnsizedType.lub_ad_type` really works for `TupleAD`, for example, this `Fatal error`s:
```stan
parameters {
  real x;
}
model {
  array[2] tuple(real, real) z = {(1,x),(x,1)};
}
```

Reply:
Made a pass at fixing `lub_ad_type`